src/Components/Analyzers/src/ComponentParameterUsageAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Components.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComponentParameterUsageAnalyzer : DiagnosticAnalyzer
+    {
+        public ComponentParameterUsageAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                DiagnosticDescriptors.ComponentParametersShouldNotBeSetOutsideOfComponents,
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!ComponentSymbols.TryCreate(context.Compilation, out var symbols))
+                {
+                    // Types we need are not defined.
+                    return;
+                }
+
+                context.RegisterOperationBlockStartAction(startBlockContext =>
+                {
+                    var assignedPropertyReferences = new List<IPropertyReferenceOperation>();
+                    startBlockContext.RegisterOperationAction(context =>
+                    {
+                        foreach (var operation in context.Operation.Children)
+                        {
+                            if (operation.Kind == OperationKind.PropertyReference)
+                            {
+                                assignedPropertyReferences.Add((IPropertyReferenceOperation)operation);
+                            }
+                        }
+                    }, OperationKind.SimpleAssignment, OperationKind.CompoundAssignment, OperationKind.CoalesceAssignment);
+
+                    startBlockContext.RegisterOperationBlockEndAction(context =>
+                    {
+                        for (var i = 0; i < assignedPropertyReferences.Count; i++)
+                        {
+                            var propertyReference = assignedPropertyReferences[i];
+                            var componentProperty = propertyReference.Member;
+                            var propertyAttributes = componentProperty.GetAttributes();
+
+                            if (!ComponentFacts.HasParameterAttribute(symbols, propertyAttributes))
+                            {
+                                // This is not a property reference that we care about, it is not decorated with [Parameter].
+                                continue;
+                            }
+
+                            var propertyLocationContainingType = componentProperty.ContainingType;
```

Comment:
It's a little strange to see `Location` as part of this variable name no? It doesn't have to do with locations in source code, it has to do with the containing type. Saying location here, makes me think it relates to the call site (where the assignment was, but it actually **super isn't that**).